evm/src/cpu/dup_swap.rs

```diff
@@ -0,0 +1,249 @@
+        let constr = builder.mul_extension(filter, diff);
+        yield_constr.constraint(builder, constr);
+    }
+}
+
+/// Set `used`, `is_read`, and address for channel.
+///
+/// `offset` is the stack index before this instruction is executed, e.g. `0` for the top of the
+/// stack.
+pub fn constrain_channel_packed<P: PackedField>(
+    is_read: bool,
+    filter: P,
+    offset: P,
+    channel: &MemoryChannelView<P>,
+    lv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    yield_constr.constraint(filter * (channel.used - P::ONES));
+    yield_constr.constraint(filter * (channel.is_read - P::Scalar::from_bool(is_read)));
+    yield_constr.constraint(filter * (channel.addr_context - lv.context));
+    yield_constr.constraint(
+        filter * (channel.addr_segment - P::Scalar::from_canonical_u64(Segment::Stack as u64)),
+    );
+    // // Top of the stack is at `addr = lv.stack_len - 1`.
+    let addr_virtual = lv.stack_len - P::ONES - offset;
+    yield_constr.constraint(filter * (channel.addr_virtual - addr_virtual));
+}
+
+/// Set `used`, `is_read`, and address for channel.
+///
+/// `offset` is the stack index before this instruction is executed, e.g. `0` for the top of the
+/// stack.
+pub fn constrain_channel_ext_circuit<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut CircuitBuilder<F, D>,
+    is_read: bool,
+    filter: ExtensionTarget<D>,
+    offset: ExtensionTarget<D>,
+    channel: &MemoryChannelView<ExtensionTarget<D>>,
+    lv: &CpuColumnsView<ExtensionTarget<D>>,
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+) {
+    {
+        let constr = builder.mul_sub_extension(filter, channel.used, filter);
+        yield_constr.constraint(builder, constr);
+    }
+    {
+        let constr = if is_read {
+            builder.mul_sub_extension(filter, channel.is_read, filter)
+        } else {
+            builder.mul_extension(filter, channel.is_read)
+        };
+        yield_constr.constraint(builder, constr);
+    }
+    {
+        let diff = builder.sub_extension(channel.addr_context, lv.context);
+        let constr = builder.mul_extension(filter, diff);
+        yield_constr.constraint(builder, constr);
+    }
+    {
+        let constr = builder.arithmetic_extension(
+            F::ONE,
+            -F::from_canonical_u64(Segment::Stack as u64),
+            filter,
+            channel.addr_segment,
+            filter,
+        );
+        yield_constr.constraint(builder, constr);
+    }
+    {
+        let constr = builder.add_extension(channel.addr_virtual, offset);
+        let constr = builder.sub_extension(constr, lv.stack_len);
+        let constr = builder.mul_add_extension(filter, constr, filter);
+        yield_constr.constraint(builder, constr);
+    }
+}
+
+/// Disable unused channels (everything except the first `used_in` and the last `used_out`).
+pub fn disable_unused_channels_packed<P: PackedField>(
```

Comment:
This will be in `stack.rs` later right, once `disable_other_channels` is replaced with something more flexible? I think we shouldn't add any more code like this, since it will be replaced soon, we can just leave a TODO in stack.rs.

Reply:
Sure! I’m not exactly clear on how to do that nicely, but we’ll figure something out, and I’ll delete these constraints in the meantime.